python/ray/data/_internal/util.py

```diff
@@ -37,28 +44,35 @@ def _lazy_import_pyarrow_dataset() -> LazyModule:
 
 def _check_pyarrow_version():
     global _VERSION_VALIDATED
+
     if not _VERSION_VALIDATED:
-        import pkg_resources
+        if os.environ.get(RAY_DISABLE_PYARROW_VERSION_CHECK, "0") == "1":
+            _VERSION_VALIDATED = True
+            return
+
+        from pkg_resources import require, packaging, DistributionNotFound
 
         try:
-            version_info = pkg_resources.require("pyarrow")
-            version_str = version_info[0].version
-            version = tuple(int(n) for n in version_str.split(".") if "dev" not in n)
-            if version < MIN_PYARROW_VERSION:
+            version_info = require("pyarrow")
+            version = version_info[0].version
+            if (
+                packaging.version.parse(version)
+                < packaging.version.parse(MIN_PYARROW_VERSION)
+            ) or (
+                packaging.version.parse(version)
+                >= packaging.version.parse(MAX_PYARROW_VERSION)
+            ):
                 raise ImportError(
-                    "Datasets requires pyarrow >= "
-                    f"{'.'.join(str(n) for n in MIN_PYARROW_VERSION)}, "
-                    f"but {version_str} is installed. Upgrade with "
-                    "`pip install -U pyarrow`."
+                    f"Datasets requires pyarrow >= {MIN_PYARROW_VERSION}, < "
+                    f"{MAX_PYARROW_VERSION}, but {version} is installed. Upgrade with "
+                    f"`pip install -U pyarrow<{MAX_PYARROW_VERSION}`."
```

Comment:
Can we include in the message that the user can override this by setting `RAY_DISABLE_PYARROW_VERSION_CHECK`?

Reply:
@matthewdeng Good point! I'll add that.